core/arch/arm/mm/tee_mmu.c

```diff
@@ -317,40 +317,86 @@ TEE_Result vm_map(struct user_ta_ctx *utc, vaddr_t *va, size_t len,
 	return res;
 }
 
-TEE_Result vm_set_prot(struct user_ta_ctx *utc, vaddr_t va, size_t len,
-		       uint32_t prot)
+static TEE_Result find_exact_vm_region(struct user_ta_ctx *utc, vaddr_t va,
+				       size_t len, struct vm_region **r_ret)
 {
-	struct vm_region *r;
+	struct vm_region *r = NULL;
 
-	/*
-	 * To keep thing simple: specified va and len has to match exactly
-	 * with an already registered region.
-	 */
 	TAILQ_FOREACH(r, &utc->vm_info->regions, link) {
 		if (core_is_buffer_intersect(r->va, r->size, va, len)) {
 			if (r->va != va || r->size != len)
 				return TEE_ERROR_BAD_PARAMETERS;
-			if ((r->attr & TEE_MATTR_PROT_MASK) == prot)
-				return TEE_SUCCESS;
-			if (mobj_is_paged(r->mobj)) {
-				if (!tee_pager_set_uta_area_attr(utc, va, len,
-								 prot))
-					return TEE_ERROR_GENERIC;
-			} else if ((prot & TEE_MATTR_UX) &&
-				   (r->attr & (TEE_MATTR_UW | TEE_MATTR_PW))) {
-				cache_op_inner(DCACHE_AREA_CLEAN,
-					       (void *)va, len);
-				cache_op_inner(ICACHE_INVALIDATE, NULL, 0);
-			}
-			r->attr &= ~TEE_MATTR_PROT_MASK;
-			r->attr |= prot & TEE_MATTR_PROT_MASK;
+
+			*r_ret = r;
 			return TEE_SUCCESS;
 		}
 	}
 
 	return TEE_ERROR_ITEM_NOT_FOUND;
 }
 
+TEE_Result vm_set_prot(struct user_ta_ctx *utc, vaddr_t va, size_t len,
+		       uint32_t prot)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct vm_region *r = NULL;
+
+	/*
+	 * To keep thing simple: specified va and len has to match exactly
+	 * with an already registered region.
+	 */
+	res = find_exact_vm_region(utc, va, len, &r);
+	if (res)
+		return res;
+
+	if ((r->attr & TEE_MATTR_PROT_MASK) == prot)
+		return TEE_SUCCESS;
+
+	if (mobj_is_paged(r->mobj)) {
+		if (!tee_pager_set_uta_area_attr(utc, va, len,
+						 prot))
+			return TEE_ERROR_GENERIC;
+	} else if ((prot & TEE_MATTR_UX) &&
+		   (r->attr & (TEE_MATTR_UW | TEE_MATTR_PW))) {
+		cache_op_inner(DCACHE_AREA_CLEAN,
+			       (void *)va, len);
+		cache_op_inner(ICACHE_INVALIDATE, NULL, 0);
+	}
+
+	r->attr &= ~TEE_MATTR_PROT_MASK;
+	r->attr |= prot & TEE_MATTR_PROT_MASK;
+
+	return TEE_SUCCESS;
+}
+
+static void umap_remove_region(struct vm_info *vmi, struct vm_region *reg)
+{
+	TAILQ_REMOVE(&vmi->regions, reg, link);
+	free(reg);
+}
+
+TEE_Result vm_unmap(struct user_ta_ctx *utc, vaddr_t va, size_t len)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct vm_region *r = NULL;
+
+	/*
+	 * To keep thing simple: specified va and len has to match exactly
+	 * with an already registered region.
+	 */
+	res = find_exact_vm_region(utc, va, ROUNDUP(len, SMALL_PAGE_SIZE), &r);
+	if (res)
+		return res;
+
+	if (mobj_is_paged(r->mobj)) {
+		tee_pager_rem_uta_region(utc, r->va, r->size);
+		maybe_free_pgt(utc, r);
```

Comment:
`maybe_free_pgt()` needs to be called for unpaged regions at other places.

Reply:
I'm not sure I understand. You mean I should move this call outside the block to always be called regardless if it's paged or not?

Reply:
I think so, yes, as in `tee_mmu_rem_rwmem()` and `tee_mmu_clean_param()`. Am I wrong?

Reply:
You're right, well spotted.